# Loops

In Python, we have two structures for loops: the for loop and the while loop (there is no do-while loop).


It is important to have a clear understanding of the concepts of a counter and an accumulator, as they are essential when working with any type of loop.

A counter is a variable that keeps track of the number of times a particular event or condition occurs. It is commonly used in loops to control the iteration process. The counter is usually initialized before the loop starts and is incremented or decremented with each iteration.

An accumulator, on the other hand, is a variable that accumulates or collects values over multiple iterations of a loop. It is often used to calculate a sum, a product, or any other aggregate value based on the values obtained during each iteration of the loop. The accumulator is typically initialized before the loop and updated within the loop body.

Both counters and accumulators play crucial roles in controlling and managing the flow of a loop, allowing you to perform specific actions based on the current state or progress of the loop iteration.

# While


They are also called indeterminate loops, and they are used when we know the condition that must be met to end the loop, but we do not know how many times or how long it will take.

In [ ]:

while condition:
  #Instructions inside the loop

#Instructions outside the loop

The structure of the loop is very simple, as the only thing that needs to be specified is the condition that must be met to continue in the loop. Essentially, it functions similarly to a conditional if() statement, except that the code inside the loop will be executed as many times as necessary as long as the condition is true.

Since the loop can be understood as a conditional statement, it works in the same way for boolean values, as well as for the values 1 and 0 explained in other materials.

In [ ]:

a = 1
while a:
  print("Choose an option:\n1. ---\n2. ---\n0. Exit\n\n")
  a = int(input())


Choose an option:
1. ---
2. ---
0. Exit


1
Choose an option:
1. ---
2. ---
0. Exit


0



In the previous example, we relied on the value of 'a' itself as the condition for the loop. However, assigning a value beforehand can sometimes cause issues. Another way to do it is by directly performing a comparison as the loop condition.

In [ ]:
number = int(input('Enter an integer number. 0 to finish: '))

while number != 0:
number = int(input('Enter an integer number. 0 to finish: '))

print('End of the program.')

Then, it is possible to use infinite loops like while(True) or while(1) if something is required to run all the time.\
Though it is not ideal to use it with the while() loop, it exists the statement break, which allows to get out of the loop in the previous scenaries, in case that it is critical not to run the remaining lines before ending the loop.

A common error using while() loops consists of not updating the control variable, thus creating a non intentional infinite loop (a bug in the code).

In [ ]:
""" (Do not execute this cell)
i = 5

# Run this loop as long as i is less than 15
while i < 15:
    # Show a message
    print("¡Hello world!")

"""

' (No ejecutar esta celda sin estar comentada)\ni = 5\n\n# Ejecutar este ciclo mientras i es menor que 15\nwhile i < 15:\n    # Mostrar un mensaje\n    print("¡Hola mundo!")\n\n'

## Exercise 1
Now, an example where we use counters and accumulators.
The user is asked to enter an integer number every time, and the process will be repeated with more numbers until the user enters a zero. At the end (when the zero is entered), show the amount of numbers entered, the total sum of those number and their average value.

## Exercise 2
Use the input() function to ask for the user to enter one integer number. Only if the number is even, it will be accumulated. The program must come to an end if the sum of the numbers surpasses 50 or if the user enters a maximum of 10 numbers.

# For

for() loops are used to have complete control of how many iterations will be done. It is necessary to have an "initial" value and a "final" value. Optionally we can have a determined step size.\
Opposite to C, for example, where we have all of these arguments in the loop statement, in Python we have to use the [range](https://www.w3schools.com/python/ref_func_range.asp) function (Important: Python takes until the second to last element).

In [ ]:
for counter_variable in range(initial_value, final_value, step_size):
    ...
    ...
    # Bloque de instrucciones dentro del ciclo...
    ...
    ...

## Bloque de instrucciones fuera del ciclo...

For example, it we want a code to show all the even numbers from 900 to 1000, we have:

In [ ]:
for i in range(900, 1000, 2):
    print(i)

900
902
904
906
908
910
912
914
916
918
920
922
924
926
928
930
932
934
936
938
940
942
944
946
948
950
952
954
956
958
960
962
964
966
968
970
972
974
976
978
980
982
984
986
988
990
992
994
996
998


We can see how it printed until 998 (as it was shown, we take until one number prior; if we would like to include 1000, it should be at least until 1001).


We can take advantage of the default arguments of the range() function for the following example, where we take the numbers from 0 to 10000 and we count all of them which are multiple of 33.

In [ ]:
counter = 0 # Iniciamos el contador en cero

for i in range(10000):
    if (i % 33 == 0): # Preguntamos si el residuo es 0 (es múltiplo de 33)
        counter += 1 # Si es múltiplo aumentamos el contador en 1

    # Si no es múltiplo no hacemos nada

#Mostramos el valor del contador
print(counter)

304


For loops are used for excellence to go over arrays, lists or matrices, as their dimensions are fixed values, and allow to iterate its elements at please.

In [ ]:
a = [8, 10, 23, 17, 15]

for i in range(len(a)):
  print(a[i])

8
10
23
17
15


It's important to note how len(a) returns the number of elements, but in Python, the index goes from zero, that's why all of the elements were printed. Let's see the difference:

In [ ]:
len(a)

5

In [ ]:
for i in range(4):
  print(a[i])

8
10
23
17


In [ ]:
for i in range(5):
  print(a[i])

8
10
23
17
15


Then, as we have nested conditionals, we can also have nested loops, where each iteration of the external loops makes one full iteration of the internal loop. Combining it with the things said before, let's see an example with a matrix (list of lists):

In [ ]:
M = [[1,3,4],[5,6,7],[8,9,2]]
M

[[1, 3, 4], [5, 6, 7], [8, 9, 2]]

In [ ]:
for i in range(3): # Rows
  for j in range(3): # Columns
    print("Row " + str(i) + ", Column " + str(j) + ": " + str(M[i][j]))

Fila 0, Columna 0: 1
Fila 0, Columna 1: 3
Fila 0, Columna 2: 4
Fila 1, Columna 0: 5
Fila 1, Columna 1: 6
Fila 1, Columna 2: 7
Fila 2, Columna 0: 8
Fila 2, Columna 1: 9
Fila 2, Columna 2: 2
